File: apps/trading-journal/backend/app/schemas/charts.py
"""
Charts schemas for price data and visualization.
"""

from __future__ import annotations

from pydantic import BaseModel, Field, ConfigDict, field_serializer
from typing import Optional, List
from datetime import datetime
from decimal import Decimal


class PriceDataPoint(BaseModel):
    """Single OHLCV data point."""
    timestamp: datetime = Field(..., description="Timestamp")
    open: Decimal = Field(..., description="Open price")
    high: Decimal = Field(..., description="High price")
    low: Decimal = Field(..., description="Low price")
    close: Decimal = Field(..., description="Close price")
    volume: Optional[int] = Field(None, ge=0, description="Volume")
    
    @field_serializer('open', 'high', 'low', 'close')
    def serialize_decimal(self, value: Decimal) -> float:
        """Serialize Decimal to float for JSON."""
        return float(value) if value is not None else None
    
    model_config = ConfigDict(
        json_encoders={
            datetime: lambda v: v.isoformat() if v else None,
        }
    )


class PriceDataResponse(BaseModel):
    """Price data response with metadata."""
    ticker: str = Field(..., description="Ticker symbol")
    timeframe: str = Field(..., description="Timeframe (1m, 5m, 15m, 1h, 1d)")
    start_date: datetime = Field(..., description="Start date")
    end_date: datetime = Field(..., description="End date")
    data: List[PriceDataPoint] = Field(..., description="Price data points")
    
    model_config = ConfigDict(
        json_encoders={
            datetime: lambda v: v.isoformat() if v else None,
        }
    )


class TradeOverlayData(BaseModel):
    """Trade overlay data for chart visualization."""
    trade_id: int = Field(..., description="Trade ID")
    ticker: str = Field(..., description="Ticker symbol")
    entry_time: datetime = Field(..., description="Entry timestamp")
    entry_price: Decimal = Field(..., description="Entry price")
    exit_time: Optional[datetime] = Field(None, description="Exit timestamp")
    exit_price: Optional[Decimal] = Field(None, description="Exit price")
    side: str = Field(..., description="Trade side (LONG or SHORT)")
    net_pnl: Optional[Decimal] = Field(None, description="Net P&L")
    
    model_config = ConfigDict(
        json_encoders={
            Decimal: lambda v: float(v) if v is not None else None,
        }
    )

File: apps/trading-journal/frontend/src/types/charts.ts
/**
 * TypeScript types for chart-related data structures.
 * 
 * These types match the Pydantic schemas from the backend.
 */

export interface PriceDataPoint {
  timestamp: string // ISO datetime string
  open: number
  high: number
  low: number
  close: number
  volume?: number | null
}

export interface PriceDataResponse {
  ticker: string
  timeframe: string
  start_date: string // ISO datetime string
  end_date: string // ISO datetime string
  data: PriceDataPoint[]
}

export interface TradeOverlayData {
  trade_id: number
  ticker: string
  entry_time: string // ISO datetime string
  entry_price: number
  exit_time?: string | null // ISO datetime string
  exit_price?: number | null
  side: string
  net_pnl?: number | null
}

export type Timeframe = '1m' | '5m' | '15m' | '1h' | '1d'
export type ChartMode = 'candlestick' | 'line'

export interface ChartColorConfig {
  // Chart background and layout
  background?: string
  textColor?: string
  gridLines?: string
  borderColor?: string
  
  // Candlestick colors
  upColor?: string
  downColor?: string
  wickUpColor?: string
  wickDownColor?: string
  
  // Line chart color
  lineColor?: string
  
  // Indicator colors
  volumeColor?: string
  movingAverageColor?: string
  movingAverageColor2?: string
}

export interface ChartIndicatorConfig {
  // Moving averages
  showSMA?: boolean
  smaPeriod?: number
  showEMA?: boolean
  emaPeriod?: number
  
  // Volume
  showVolume?: boolean
  
  // Other indicators (can be added later)
  showRSI?: boolean
  rsiPeriod?: number
}

